File: ReverseLinkedList.java
/**
 * Created by yhf on 15/5/5.
 */

/*
Reverse a singly linked list.

click to show more hints.

Hint:
A linked list can be reversed either iteratively or recursively. Could you implement both?
*/

public class ReverseLinkedList {
    public static ListNode reverseList(ListNode head) {
        // iteratively
        if (head == null || head.next == null) {
            return head;
        }
        ListNode slow = head, fast = head.next;
        while (fast != null && fast.next != null) {
            ListNode temp = fast.next;
            fast.next = slow;
            slow = fast;
            fast = temp;
        }
        fast.next = slow;
        head.next = null;
        return fast;
    }

    public static ListNode reverseList2(ListNode head) {
        // recursively
        if (head == null || head.next == null) {
            return head;
        } else {
            ListNode newHead = reverseList2(head.next);
            head.next.next = head;
            head.next = null;
            return newHead;
        }
    }

    public static void main(String[] args) {
        ListNode n1 = new ListNode(1);
        ListNode n2 = new ListNode(2);
        ListNode n3 = new ListNode(3);
        ListNode n4 = new ListNode(4);
        ListNode n5 = new ListNode(5);
        ListNode n6 = new ListNode(6);
        n1.next = n2;
        n2.next = n3;
        n3.next = n4;
        n4.next = n5;
        n5.next = n6;
        ListNode head = reverseList2(n1);
    }
}

File: ReverseLinkedList.py
"""
Reverse a singly linked list.

Hint:
A linked list can be reversed either iteratively or recursively. Could you implement both?
"""
from ListNode import ListNode


class Solution:
    # @param {ListNode} head
    # @return {ListNode}
    def reverseList(self, head):
        # iteratively
        if not head or not head.next:
            return head
        pre = head
        cur = head.next
        while cur:
            temp = cur.next
            cur.next = pre
            pre = cur
            cur = temp
        head.next = None
        return pre

    def reverseListRecursively(self, head):
        # recursively
        if not head or not head.next:
            return head
        new_head = self.reverseListRecursively(head.next)
        head.next.next = head
        head.next = None
        return new_head


if __name__ == '__main__':
    n1 = ListNode(1)
    n2 = ListNode(2)
    n3 = ListNode(3)
    n4 = ListNode(4)
    n5 = ListNode(5)

    n1.next = n2
    n2.next = n3
    n3.next = n4
    n4.next = n5

    head = Solution().reverseListRecursively(n1)
